src/main/scala/chisel3/util/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,234 @@
+        module
+      case Some(externalModule: DefModule) =>
+        externalModule
+      case _ =>
+        throw new Exception(s"Could not find module $moduleName in circuit $circuit")
+    }
+  }
+
+  /**
+    * walk the module and for memories that have LoadMemory annotations
+    * generate the bindable modules for verilog emission
+    *
+    * @param modulePrefix kind of a path to the current module
+    * @param myModule     module being searched for memories
+    */
+  def processModule(modulePrefix: String, myModule: DefModule): Unit = {
+
+    def makePath(componentName: String): String = {
+      circuitState.circuit.main + "." + myModule.name + "." + componentName
+    }
+
+    def processMemory(name: String): Unit = {
+      val fullMemoryName = makePath(s"$name")
+
+      memoryAnnotations.find {
+        ma: LoadMemoryAnnotation =>
+          val targetName = ma.target.serialize
+          targetName == fullMemoryName
+      } match {
+        case Some(lma @ LoadMemoryAnnotation(
+        ComponentName(componentName, moduleName: ModuleName), _, hexOrBinary, _
+        )) =>
+
+          val writer = new java.io.StringWriter
+          circuitState.circuit.modules
+            .filter { module => module.name == moduleName.name }
+            .collectFirst { case m: firrtl.ir.Module => m }
+            .foreach { module =>
+
+              val moduleMap = circuitState.circuit.modules.map(m => m.name -> m).toMap
+              val renderer = verilogEmitter.getRenderer(module, moduleMap)(writer)
+              val loadFileName = lma.getFileName + ".txt"
+              renderer.emitVerilogBind(s"BindsTo_${moduleName.name}",
+                s"""
+                   |initial begin
+                   |  $$readmem$hexOrBinary("$loadFileName", ${myModule.name}.$componentName);
+                   |end
+                    """.stripMargin)
+              val inLineText = writer.toString + "\n" +
+                s"""bind ${myModule.name} BindsTo_${myModule.name} BindsTo_${myModule.name}_Inst(.*);"""
+
+              val blackBoxInline = BlackBoxInlineAnno(
+                moduleName,
+                moduleName.serialize + "." + componentName + ".v",
+                inLineText
+              )
+
+              bindModules += blackBoxInline
+            }
+
+        case _ =>
+      }
+    }
+
+    def processStatements(statement: Statement): Unit = {
+      statement match {
+        case block: Block =>
+          block.stmts.foreach { subStatement =>
+            processStatements(subStatement)
+          }
+
+        case m: DefAnnotatedMemory => processMemory(m.name)
+
+        case m: DefMemory          => processMemory(m.name)
+
+        case WDefInstance(_, _, moduleName, _) =>
+          val subModule = findModule(moduleName, circuitState.circuit)
+          val newPrefix = (if (modulePrefix.nonEmpty) modulePrefix + "." else "") + myModule.name
+
+          processModule(newPrefix, subModule)
```

Comment:
Without instance annotations, I'm not convinced that this PR needs to do an instance walk (and if it did, I'd expect to see use of `InstanceGraph`). The expressiveness of a `LoadMemoryAnnotation` is limited to components. I think this PR can just loop over all modules that have annotations targeting their components. An explicit instance walk would seem to visit multiple modules multiple times which, without instance annotations, should be unnecessary.

This brings up a larger problem of how a user can use this to initialize different memories differently, e.g., the boot ROMs of disparate Rocket cores. Likely out of scope without a good formalism for instance annotations.